Fix: Hang when `ElectrumBlockchainConfig::stop_gap == 0`

evanlinjin opened the pull request:
* Ensure `chunk_size` is > 0 during wallet sync.

* Slight refactoring for better readability.

* Add test: `test_electrum_blockchain_factory_sync_with_stop_gaps`

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Description

`Wallet::sync` hangs indefinitely when syncing with Electrum with `stop_gap` set as 0.

The culprit is having `chunk_size` set as `stop_gap`. A zero value results in syncing not being able to progress.

Fixes #651

### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### Bugfixes:

~* [ ] This pull request breaks the existing API~
* [x] I've added tests to reproduce the issue which are now passing
* [x] I'm linking the issue being fixed by this PR

evanlinjin committed:
Fix hang when `ElectrumBlockchainConfig::stop_gap == 0`

* Ensure chunk_size is > 0 during wallet sync.
* Slight refactoring for better readability.
* Add test: test_electrum_blockchain_factory_sync_with_stop_gaps

evanlinjin commented:
I realized for the test, I'm using a single key descriptor, when I really should be using an extended key, since we are testing the stop gap.

So this is still WIP.

LLFourn commented on src/blockchain/electrum.rs:
```diff
@@ -385,4 +386,116 @@ mod test {
     }
+
+    #[test]
+    fn test_electrum_blockchain_factory_sync_with_stop_gaps() {
```
can we put this in `blockchain_tests` so it covers all blockchains?

evanlinjin replied:
That's a good idea, I'll do so in a bit.

evanlinjin replied:
> can we put this in `blockchain_tests` so it covers all blockchains?

I feel like this is too much work for this PR. Should we get this merged and I can create a second ticket?

Each test case requires creating a new `Blockchain` with a specified `stop_gap`.

I am also not very familiar with macros in Rust, so this may take a while.

LLFourn replied:
Hmmmm. Ok I see and agree with this point. Not all blockchains have "stop gaps" anyways.

evanlinjin replied:
We can definitely do something universal I think. Can create a custom test trait with a method such as `generate_blockchain_with_stop_gap` and have that trait as an input.

evanlinjin replied:
I've done something relatively universal, we can continue this conversation here: https://github.com/bitcoindevkit/bdk/pull/652#issuecomment-1173789385

LLFourn commented:
Interesting bug. I'm not sure about the semantics of `stop_gap = 0`? Maybe the correct thing is to do nothing. If I am going to stop after I find a gap of 0 then I should just stop right away no? (you start off with a gap of 0).

LLFourn commented on src/blockchain/electrum.rs:
```diff
@@ -108,7 +108,10 @@ impl WalletSync for ElectrumBlockchain {
         let mut tx_cache = TxCache::new(database, &self.client);
-        let chunk_size = self.stop_gap;
+
+        // Ensure chunk_size > 0, otherwise we will result in an infinite loop
```
I understand better what is going on now. We must find a gap greater than `stop_gap` in order to stop.
So I think this comment should be: 

> set the chunk_size to the smallest value that we can find a gap greater than stop gap.

It is also a performance issue to set it to `stop_gap` since you will do two requests when you should only do one on any empty wallet (for example).

evanlinjin replied:
Changed in 475b005eb9151284885600896e8f0f50a45cd1b2.

LLFourn commented:
ACK. See https://github.com/bitcoindevkit/bdk/pull/652#discussion_r912656091 

`chunk_size` was set wrongly before. Thanks!

afilini commented on src/blockchain/electrum.rs:
```diff
@@ -145,18 +148,16 @@ impl WalletSync for ElectrumBlockchain {
                     // collect up to chunk_size heights to fetch from electrum
                     let needs_block_height = {
-                        let mut needs_block_height_iter = conftime_req
+                        let mut needs_block_height = HashSet::with_capacity(chunk_size);
```
I think here you can just construct the `HashSet` from the iterator, using the `FromIterator` trait:

```rust
let needs_block_height = conftime_req./*<long chain>*/.collect();
```

afilini commented on src/blockchain/electrum.rs:
```diff
@@ -385,4 +386,116 @@ mod test {
+        // [1]   actual_gap: Range size of address indexes without a balance
+        // [2] addrs_before: Range size of address indexes (before gap) which contains a balance
+        // [3]  addrs_after: Range size of address indexes (after gap) which contains a balance
+        let test_vectors: Vec<[u64; 4]> = vec![
```
Wouldn't it be better to use `usize` here? I see that you have to cast these values in a few places

evanlinjin replied:
It seems that if we do this, other variables will need casting... Should we leave it as is?

afilini replied:
Yeah at a quick glance I thought using usize would be fine, but if you need to cast anyway there's no reason to change it

afilini reviewed (commented):
Concept ACK, a couple of things:

- did you check if this happens with esplora as well? The two blockchains are very similar in terms of architecture and code
- once you are done fixing the various comments, could you squash the commits into the first one?

evanlinjin committed:
`Blockchain` stop_gap testing improvements

This is a continuation of the #651 fix. We should also check whether the
same bug affects esplora as noted by @afilini. To achieve this, I've
introduced a `ConfigurableBlockchainTester` trait that can test multiple
blockchain implementations.

* Introduce `ConfigurableBlockchainTester` trait.
* Use the aforementioned trait to also test esplora.
* Change the electrum test to also use the new trait.
* Fix some complaints by clippy in ureq.rs file (why is CI not seeing
  this?).
* Refactor some code.

evanlinjin commented:
> did you check if this happens with esplora as well? The two blockchains are very similar in terms of architecture and code

@afilini I've made a trait that can be reused for testing both esplora and electrum. I've also made it expandable, so in the future we can implement more tests in it. Let me know whether these changes make sense!

evanlinjin commented:
> once you are done fixing the various comments, could you squash the commits into the first one?

@afilini I feel like it's better to keep it to two commits. One if for the actual fix, the other is introducing a new testing trait that can be shared between different blockchain implementations.

What do you think?

evanlinjin commented:
@afilini Should this fix also be part of [Release 0.20.0 Feature Freeze](https://github.com/bitcoindevkit/bdk/milestone/16)?

Ensuring that the `Blockchain` implementations respect the `stop_gap` parameter can potentially be important for avoiding address reuse.

The test I have included checks for this.

https://github.com/bitcoindevkit/bdk/blob/612da165f8cfbc2b0aee0a37b2cdc44d6da017c8/src/testutils/configurable_blockchain_tests.rs#L131-L135

afilini commented:
> @afilini I feel like it's better to keep it to two commits. One if for the actual fix, the other is introducing a new testing trait that can be shared between different blockchain implementations.

Yeah two commits works for me, as long as they are self-contained (they should both compile and run `cargo test` with no errors) and they have a good commit message

> @afilini Should this fix also be part of [Release 0.20.0 Feature Freeze](https://github.com/bitcoindevkit/bdk/milestone/16)?

I don't think this is as critical as that other one, but if this is ready in time yes, this is also important

afilini reviewed (approved):
ACK 612da165f8cfbc2b0aee0a37b2cdc44d6da017c8
